src/server/enterprise/server/api_server.go

```diff
@@ -330,3 +341,114 @@ func (a *apiServer) Deactivate(ctx context.Context, req *ec.DeactivateRequest) (
 
 	return &ec.DeactivateResponse{}, nil
 }
+
+func (a *apiServer) Pause(ctx context.Context, req *ec.PauseRequest) (resp *ec.PauseResponse, retErr error) {
+	if err := scaleDownWorkers(ctx, a.env.GetKubeClient(), a.env.Namespace); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	if err := rollPachd(ctx, a.env.GetKubeClient(), a.env.Namespace, true); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+
+	return &ec.PauseResponse{}, nil
+}
+
+func rollPachd(ctx context.Context, kc *kubernetes.Clientset, namespace string, paused bool) error {
+	dd := kc.AppsV1().Deployments(namespace)
+	d, err := dd.Get(ctx, "pachd", metav1.GetOptions{})
+	if err != nil {
+		return errors.Errorf("could not get pachd deployment: %v", err)
+	}
+	// Updating the spec rolls the deployment, killing each pod and causing
+	// a new one to start.
+	for i, c := range d.Spec.Template.Spec.Containers {
+		if len(c.Command) == 0 {
+			continue
+		}
+		if c.Command[0] == "/pachd" {
+			if paused {
+				c.Command = []string{"/pachd", "--mode", "paused"}
```

Comment:
would this interact in a weird way with helm? Wouldn't an upgrade overwrite these values? I think recording the state via a change on the container spec may not be the way to go. Would a configmap make more sense if we're using kubernetes/env as a place to store this info?

Reply:
I’ve rewritten the code to use a ConfigMap unmanaged by Helm (which thus will not be overwritten by Helm operations).